### DISCUSSION TITLE
Faster Archiving from a Windows Plotter

### DISCUSSION THREAD

**Discussion by zorner** [ericaltendorf/plotman#962] [Show and tell]
This is a duplicate discussion from [Machinaris repo](https://github.com/guydavis/machinaris/discussions/814).

**Background**

I/O operations in a virtual environment on Windows can be very slow. In particular disk I/O. This is caused by WSL version 2. WSL v2 has many improvements, but at the cost of disk I/O. Docker on Windows runs inside WSL v2. Thus plotting and archiving are very slow. Example 2.5 hours per plot using Madmax in Docker vs 35 minutes on Windows using Madmax with the same settings. Also 3+ hours to move a plot to a remote host from docker vs 25 minutes in WSL v1.

WSL v1 has better I/O access. Docker does not run on WSL v1. Instead it used hyper-visor. Machinaris in Docker using hyper-visor crashes. (May be fixable, not worth the effort currently.) 

**Setup**
Install linux VM on Windows. There are many distributions available in the MS store. Here we use Ubuntu.
```
> wsl --install -d Ubuntu
```
Ubuntu will launch into its own shell. Set username and password Then run these:
```
$ sudo apt-get update
```
```
$ sudo apt-get install curl rsync python3-venv git software-properties-common -y
```
Get Chia CLI - There is no need for the GUI.
```
$ curl -L -o chia-latest.deb https://download.chia.net/latest/x86_64-Ubuntu-cli
```
```
$ sudo dpkg -i chia-latest.deb
```
We are done in Ubuntu for now. Back to the windows command line.
```
$ exit
```
Make Ubuntu the default distribution.
```
> wsl -s Ubuntu
```
Make Ubuntu use WSL v1
```
> wsl --set-version Ubuntu 1
```
Cancel restart when asked.
Now back to Ubuntu, but as the root user. (There are files that need to accessed that only root can do, Windows will not allow the owner or group to change.) If you are not trying to sync up with Machinaris, you do not need to be the root user.
```
> wsl --user root
```
Go to where Chia is installed.
```
# cd /opt/chia
```
Create the venv.
```
#  python3 -m venv venv
```
```
#  ln -s venv/bin/activate
```
Activate the venv
```
# . ./activate
```
Install Plotman
```
(venv)# pip install --force-reinstall git+https://github.com/guydavis/plotman@development
```
```
(venv)# pip3 install pyyaml
```
```
(venv)# plotman config generate
```
Now we need to replicate some setup from Machinaris so everything is pointed to the same file inside the docker container and here in Ubuntu. Replace {USER_HOST} with the Windows username.
```
(venv)# mv /root/.config/plotman/plotman.yaml /root/.config/plotman/plotman.org.yaml
```
```
(venv)# ln -s /mnt/c/Users/{USER_HOST}/.machinaris/plotman/plotman.yaml /root/.config/plotman/plotman.yaml
```
```
(venv)# ln -s /mnt/c/Users/{USER_HOST}/.machinaris/ /root/.chia
```
```
(venv)# ln -s /opt/chia/madmax/chia_plot /opt/chia/venv/bin/
```
```
(venv)# ln -s /opt/chia/madmax/chia_plot_k34 /opt/chia/venv/bin/
```
Make symbolic links to match the volumes mounted in docker-composer.yml for tmp, tmp2, and dst directories. Yours will likely be different.
```
(venv)# ln -s /mnt/r/ /plotting1
```
```
(venv)# ln -s /mnt/s/ /plotting2
```
```
(venv)# ln -s /mnt/t/ /ramdrive
```
To not have to activate the venv every time, edit /etc/profile. At the end of the file add 'export PATH="/opt/chia/venv/bin:$PATH" '  There are other method to update $PATH.
```
(venv)# nano /etc/profile
```
![image](https://user-images.githubusercontent.com/5413793/202285451-fce0f310-1664-4e5b-8e7f-2a23872a2522.png)
Ctrl-X and Y to exit and save.

We need to re-login for the last step to take effect.
```
(venv)# deactive
```
```
# exit
```
```
> wsl --user root
```
Finally we are ready start archiving
```
# plotman archive
```

**To Upgrade pyhton to 3.9 or 3.10**
```
# add-apt-repository ppa:deadsnakes/ppa
```
Press Enter when asked
```
# apt-get install python3.10 python3.10-venv
# cd /opt/chia
# . ./activate
(venv)# python --version
(venv)# pip freeze > requirements.txt
(venv)# deactivate
# rm -R venv/*
# python3.10 -m venv venv
# . ./activate
(venv)# pyhton --version
(venv)# pip install --force-reinstall git+https://github.com/guydavis/plotman@development
(venv)# ln -s /opt/chia/madmax/chia_plot /opt/chia/venv/bin/
(venv)# ln -s /opt/chia/madmax/chia_plot_k34 /opt/chia/venv/bin/
(venv)# pip install -r requirements.txt
(venv)# deactive
# pyhton --version
```

**Comment by altendky** [comment]
In general, I would like to have plotman support Windows.  I started some work in that direction last year, but didn't complete it.  No ETA.  But, afaik you can get good performance in WSL v2, much better than v1 I thought.  https://github.com/ericaltendorf/plotman/issues/108#issuecomment-822151350

I haven't worked in WSL, it came about right at the tail end of my work-forced Windows usage, but the `sudo` for creating the env etc strike me as a bad practice.

**Reply by zorner** [reply]
[#108 (commnet)](https://github.com/ericaltendorf/plotman/issues/108#issuecomment-822151350) was about mounting drive in Linux.

Yes WSL v2 is better than WSL v1 in many ways, but at the cost of disk IO. [Comparing WSL 1 and WSL 2](https://learn.microsoft.com/en-us/windows/wsl/compare-versions) has the details on an MS site.

And yes, needing to be the root user to run all this is bad practice, but for now it is the only way I can get it to work. The issue is Machinaris only has the root user inside the container. Windows knows Machinaris is the source of certain directories and files. As such it will not let the Ubuntu VM do a chgrp to allow permissions to a local user. May be I can do something with the 'other' permissions.

Anyway, I need access to some of the files that are root/root, and being root is the only way at the moment. Also some are in .dirs adding to the permissions issues.

If your not trying to sync this with Machinaris, then absolutely drop a lot of the sym links and do venv as a normal user. 

Updated the 1st comment, as it does not make sense to do a sudo when you are the root user.

**Reply by altendky** [reply]
The suggestion was to do disk stuff the quick way instead of the slow way.  But yes, if you are trying to mash Linux-VM/WSL/Windows and run stuff in each f them separately and integrated together, there is likely to be issues.  Maybe the tmp disk could just be dedicated to WSL, mounted directly using the quick option, etc.  Of course, just going all Linux is generally the easy approach if possible.  I understand that sometimes it isn't.
